cmake/scripts/run_test.sh

```diff
@@ -18,143 +18,146 @@
-
-save () {
-  for i do printf %s\\n "$i" | sed "s/'/'\\\\''/g;1s/^/'/;\$s/\$/' \\\\/" ;
-  done
-  echo " "
-}
-
-RUN_COMMAND=$(save "$@")
+shift 2
 
 # Ensure uniform sorting for pathname expansion
 export LC_ALL=C
 
-#
-# Add a top level target to run and compare the test:
-#
-
-run(){
-  rm -f failing_output
-  rm -f output
-  rm -f stdout
-
-  eval "set -- ${RUN_COMMAND}"
-  "$@" > stdout 2>&1
-  RETURN_VALUE=$?
-
-  [ -f output ] || mv stdout output
-
-  if [ $RETURN_VALUE -ne 0 ]; then
-    mv output failing_output
-    echo "${TEST_FULL}: BUILD successful."
-    echo "${TEST_FULL}: RUN failed. ------ Return code $RETURN_VALUE"
-    echo "${TEST_FULL}: RUN failed. ------ Result: `pwd`/failing_output"
-    echo "${TEST_FULL}: RUN failed. ------ Partial output:"
-    cat failing_output
-    if [ -f stdout ]; then
-      echo ""
-      echo "${TEST_FULL}: RUN failed. ------ Additional output on stdout/stderr:"
-      echo ""
-      cat stdout
+case $STAGE in
+  run)
+    ##
+    # run stage:
+    #   - run specified command and parameters given by $@
+    #   - creates file "output" on success containing stdout and stderr of
+    #     the test
+    #   - if test exits with non-zero return value, output is renamed to
+    #     failing_output
+    ##
+
+    rm -f failing_output
+    rm -f output
+    rm -f stdout
+
+    "$@" > stdout 2>&1
+    RETURN_VALUE=$?
+
+    [ -f output ] || mv stdout output
+
+    if [ $RETURN_VALUE -ne 0 ]; then
+      mv output failing_output
+      echo "${TEST_FULL}: BUILD successful."
+      echo "${TEST_FULL}: RUN failed. ------ Return code $RETURN_VALUE"
+      echo "${TEST_FULL}: RUN failed. ------ Result: `pwd`/failing_output"
+      echo "${TEST_FULL}: RUN failed. ------ Partial output:"
+      cat failing_output
+      if [ -f stdout ]; then
+        echo ""
+        echo "${TEST_FULL}: RUN failed. ------ Additional output on stdout/stderr:"
+        echo ""
+        cat stdout
+      fi
+      exit 1
     fi
-    exit 1
-  fi
-}
 
-diff() {
-  rm -f failing_diff*
-  rm -f diff*
-  touch diff
+    ;;
+
+  diff)
+
+    ##
+    # diff stage:
+    #   - compares the file "output" against various comparison files.
+    ##
+
+    NUMDIFF_EXECUTABLE="$1"
+    DIFF_EXECUTABLE="$2"
+    COMPARISON_FILE="$3"
 
-  test_successful=false
+    rm -f failing_diff*
+    rm -f diff*
+    touch diff
 
-  #
-  # Pick up main comparison file and all variants. A valid variant name is
-  # of the form [...].output.[STRING]
-  #
-  for file in "${COMPARISON_FILE}"*; do
-    # determine variant name (empty string for main comparison file):
-    variant="${file#*.output}"
+    test_successful=false
 
     #
-    # Run diff or numdiff (if available) to determine whether files are the
-    # same. Create a diff file "diff${variant}" for each variant file that
-    # is found (including the main comparison file).
+    # Pick up main comparison file and all variants. A valid variant name is
+    # of the form [...].output.[STRING]
```

Comment:
Now, I also remember the reason. The old variant was
```
for files in "${COMPARISON_FILE}" "${COMPARISON_FILE}".*; ...
```
This produces at least a warning if the second expression doesn't match anything.